### PR TITLE
MPTokenIssuanceID New Construct and Serialization Format

### PR DESCRIPTION
**Pull request by shawnxie999**

### CODE REVIEW

File: XLS-0033d-compact-fungible-tokens/README.md [hunk 1]
```diff
@@ -67,11 +67,28 @@ The **`CFTokenIssuance`** object represents a single CFT issuance and holds data
...
+
+The ID of a `CFTokenIssuance` object, a.k.a. `CFTokenIssuanceID`, is a 192-bit integer, concatenated in order:
+
 * The transaction sequence number.
+* The AccountID of the issuer.
+
+```
+┌──────────────────────────┐┌──────────────────────────┐
+│                          ││                          │
+│      Sequence            ││     Issuer AccountID     │
```
Comment (sappenin) [inline XLS-0033d-compact-fungible-tokens/README.md:85]: Let's update the FAQ here to discuss why we're using a `Sequence` number at all -- e.g., so that an MPT can't be recreated after having been deleted (and other reasons)?

Comment (sappenin) [reply XLS-0033d-compact-fungible-tokens/README.md]: Sounds good, thanks for adding!

**Review by sappenin — APPROVED** [review]

**Review by tequdev — COMMENTED** [review]
It is not directly related to this PR change, but in this section, I believe `CurMintedAmount` is `OutstandingAmount`?

```
###### 1.2.1.1.4.3. Removing a **`MPTokenIssuance`** object
A **`MPTokenIssuance`** can be removed using the same approach, but only if the **`CurMintedAmount`** is equal to 0.
```

And section number 2.2.3.1 is duplicated

**Comment by shawnxie999** [comment]
@tequdev Good catches! Appreciate it ❤️

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 2]
```diff
@@ -907,12 +929,6 @@ Issuer also has the ability to de-authorize a holder. In that case, if the holde
 
 The original intent of the `MPTokenNode` object is that it would be a sort of "directory" (i.e., an index) that stores a list of `MPTokenID` values (each 32 bytes) that exist for a single `MPTokenIssuance`. This would allow rippled to contain an RPC endpoint that could return a paged collection of `MPToken` objects for a given issuance, or somethign similar like an RPC called `mpt_holder_balances`. In theory, this could also enable rippled to operate a sort of "clean-up" operation that could remove dangling MPTokens that still live on a ledger after a corresponding `MPTokenIssuance` has been deleted (and thus return ledger reserves back to token holders).
 
-#### 2.2.3.1 Should We Have `MPTokenNode` Directories?
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:910]: Why is this section being removed?

Comment (shawnxie999) [reply XLS-0033d-multi-purpose-tokens/README.md]: Oops my bad. Just added it back.

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 3]
```diff
@@ -67,11 +67,28 @@ The **`MPTokenIssuance`** object represents a single MPT issuance and holds data
...
+```
+┌──────────────────────────┐┌──────────────────────────┐
+│                          ││                          │
+│      Sequence            ││     Issuer AccountID     │
+│      (32 bits)           ││        (160 bits)        │
+│                          ││                          │
+└──────────────────────────┘└──────────────────────────┘
+```
+
+**Note: The `MPTokenIssuanceID` is what the user inputs for the interface. Internally, the ledger splits the `MPTokenIssuanceID` into two components: `sequence` and `issuer` address.**
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:91]: > **Note: The `MPTokenIssuanceID` is what the user inputs for the interface.

Let's clarify this a bit more. E.g., which interface are we referring to here? I think there's a difference between API "interfaces" vs. the contents of a signed transaction, and on this point these proposed changes are a bit ambiguous.

**Review by sappenin — CHANGES_REQUESTED** [review]

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 4]
```diff
@@ -963,15 +985,15 @@ This section attempts to catalog expected size, in bytes, for both Trustlines an
 |        FIELD NAME | SIZE (BITS) | SIZE (BYTES) | NOTE                                                                                                                                                       |
 |------------------:|:-----------:|:------------:|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
 |   LedgerEntryType |     16      |      2       |                                                                                                                                                            |
-| MPTokenIssuanceID |     256     |      32      |                                                                                                                                                            |
+| MPTokenIssuanceID |     192     |      32      |                                                                                                                                                            |
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:988]:
```suggestion
| MPTokenIssuanceID |     192     |      24      |                                                                                                                                                            |
```

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 5]
```diff
@@ -963,15 +985,15 @@ This section attempts to catalog expected size, in bytes, for both Trustlines an
...
 | PreviousTxnLgrSeq |     32      |      4       |                                                                                                                                                            |
 |  Field+Type Codes |     112     |      14      | For every field, there is a `FieldCode` and a `TypeCode`, taking 2 bytes in total (e.g., if there are 4 fields, we'll use 8 bytes). Here we have 7 fields. |
 |               --- |     --      |     ---      |                                                                                                                                                            |
-|             TOTAL |     832     |     104      |                                                                                                                                                            |
+|             TOTAL |     768     |     104      |                                                                                                                                                            |
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:996]:
```suggestion
|             TOTAL |     768     |     96      |                                                                                                                                                            |
```

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 6]
```diff
@@ -983,3 +1005,68 @@ As can be seen from the following size comparison table, Trustlines take up appr
 | Bytes for holding 10 Tokens |         1         |       1,450       |            2             |          3,260          |             2.2x              |
 | Bytes for holding 32 Tokens |         2         |       5,556       |            4             |         12,264          |             2.2x              |
 | Bytes for holding 64 Tokens |         3         |      13,070       |            6             |         28,444          |             2.2x              |
+
+### 2.3.2. `STAmount` serialization
+Referenced from https://gist.github.com/sappenin/2c923bb249d4e9dd153e2e5f32f96d92 with some modifications:
+
+#### Binary Encoding
+To support this idea, we first need a way to leverage the current [STAmount](https://xrpl.org/serialization.html#amount-fields) binary encoding. To accomplish this, we notice that for XRP amounts, the maximum amount of XRP (10^17 drops) only requires 57 bits. However, in the current `XRP` amount encoding, there are 62 bits available, so we can repurpose one to indicate if an amount is indeed a MPT or not.
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:1013]:
```suggestion
To support this idea, we first need a way to leverage the current [STAmount](https://xrpl.org/serialization.html#amount-fields) binary encoding. To accomplish this, we notice that for XRP amounts, the maximum amount of XRP (10^17 drops) only requires 57 bits. However, in the current `XRP` STAmount encoding, there are 62 bits available. So, so we can repurpose one of these bits to indicate if an amount is indeed a MPT or not (and still have 4 bits left over for future use, if needed). 
```

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 7]
```diff
@@ -983,3 +1005,68 @@ As can be seen from the following size comparison table, Trustlines take up appr
...
+
+The rules for reading the binary amount fields would be backward compatible, as follows:
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:1015]:
```suggestion
This enables MPT amounts to be represented in the current `STAmount` binary encoding.  he rules for reading the binary amount fields would be backward compatible, as follows:
```

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 8]
```diff
@@ -983,3 +1005,68 @@ As can be seen from the following size comparison table, Trustlines take up appr
...
+
+1. Parse off the Field ID with a type_code (`STI_AMOUNT`). This indicates the following bytes are an `STAmount`.
+2. Inspect the next bit. If `1`, then this is **not** a MPT or XRP (instead this is a regular IOU token amount). However, if the first bit is `0`, then continue.
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:1018]:
```suggestion
2. Inspect the next bit. If its value is `1`, then continue to the next step. If not, then this `STAmount` does **not** represent an MPT nor XRP (instead this is a regular IOU token amount, and can be parsed according to existing rules for those amounts). 
```

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 9]
```diff
@@ -983,3 +1005,68 @@ As can be seen from the following size comparison table, Trustlines take up appr
...
+3. Ignore (for now) the 2nd bit (this is the sign-bit, and is always 1 for both XRP and MPT).
+4. Inspect the 3rd bit. If `0`, then parse as an XRP value per usual. However, if `1`, then parse the bytes as a MPT.
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:1020]:
```suggestion
4. Inspect the 3rd bit. If `0`, then parse as an XRP value per usual. However, if `1`, then parse the remaining `STAmount` bytes as an MPT.
```

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 10]
```diff
@@ -983,3 +1005,68 @@ As can be seen from the following size comparison table, Trustlines take up appr
...
+#### Encoding for MPT Values (backward compatible)
+
+This encoding focuses on the first 3 bytes of a MPT:
+
+```
+┌────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ ┌─────────────────┐┌────────────────────────┐┌────────────┐┌─────────────────────────────────────────────┐ │
+│ │        0        ││           1            ││     1      ││                                             │ │
+│ │                 ││                        ││            ││             Remaining MPT Bytes             │ │
+│ │  "Not XRP" bit  ││  Sign bit (always 1;   ││"IsMPT" bit ││                 (389 bits)                  │ │
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:1044]:
```suggestion
│ │  "Not XRP" bit  ││  Sign bit (always 1;   ││"IsMPT" bit ││                 (256 bits)                  │ │
```

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 11]
```diff
@@ -983,3 +1005,68 @@ As can be seen from the following size comparison table, Trustlines take up appr
...
+│ │                 ││     Reserved     ││   MPT Amount Value   ││      Sequence            ││     Issuer AcountID      ││
+│ │   IsMPT=true    ││     (5 bits)     ││      (64 bits)       ││      (32 bits)           ││        (160 bits)        ││
+│ │                 ││                  ││                      ││                          ││                          ││
+│ └─────────────────┘└──────────────────┘└──────────────────────┘└──────────────────────────┘└──────────────────────────┘│
+└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+```
+
+Note: MPT introduces an extra leading byte in front of the MPT value. However, despite the MPT value being 64-bit, only 63 bits can be used. This limitation arises because internally, rippled needs to convert the value to `int64`, which has a smaller positive range compared to `uint64`.
+
+*Question*: Could the ledger perform arithmetics in `uint64`? If not, we would need to decrease the range of the MPT value.
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:1065]:
```suggestion
```

I think this question is answered. We should transition this to an FAQ answer instead.

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 12]
```diff
@@ -983,3 +1005,68 @@ As can be seen from the following size comparison table, Trustlines take up appr
...
+
+The reasons for having a construct `MPTokenIssuanceID` by concatenating `sequence` and `issuer` are:
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:1067]: This should be formulated into the FAQ. Maybe let's remove this from this PR, and then make a different PR to update the FAQ?

**Review by sappenin — CHANGES_REQUESTED** [review]

File: XLS-0033d-multi-purpose-tokens/README.md (hunk 12 above)
Comment (shawnxie999) [reply XLS-0033d-multi-purpose-tokens/README.md]: Sounds good!

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 13]
```diff
@@ -983,3 +1011,59 @@ As can be seen from the following size comparison table, Trustlines take up appr
...
+│ └─────────────────┘└────────────────────────┘└────────────┘└─────────────────────────────────────────────┘ │
+└────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+```
+
+#### Encoding for MPT Values (backward compatible)
+
+This encoding focuses on the first 3 bytes of a MPT:
+
+```
+┌────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:1046]: After talking to @shawnxie999 we agreed to adjust this section to be this:

```
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ ┌──────────────────┐ ┌──────────────────────┐ ┌──────────────────┐ ┌──────────────────┐ ┌───────────────────────────────────────────┐ │
│ |        0         | |          1           | |        0         | |                  | |                                           | │
│ |                  | |                      | |                  | |     Reserved     | |                XRP Amount                 | │
│ |  "Not XRP" bit   | | Sign bit (always 1;  | |   "IsMPT" bit    | |     (4 bits)     | |                 (57 bits)                 | │
│ | 0=XRP/MPT; 1=IOU | |      positive)       | |   0=XRP; 1=MPT   | |                  | |                                           | │
│ └──────────────────┘ └──────────────────────┘ └──────────────────┘ └──────────────────┘ └───────────────────────────────────────────┘ │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Once that change is made, I'll merge this PR.

File: XLS-0033d-multi-purpose-tokens/README.md [hunk 14]
```diff
@@ -983,3 +1011,59 @@ As can be seen from the following size comparison table, Trustlines take up appr
...
+This enables MPT amounts to be represented in the current `STAmount` binary encoding.  he rules for reading the binary amount fields would be backward compatible, as follows:
+
+1. Parse off the Field ID with a type_code (`STI_AMOUNT`). This indicates the following bytes are an `STAmount`.
+2. Inspect the next bit. If its value is `1`, then continue to the next step. If not, then this `STAmount` does **not** represent an MPT nor XRP (instead this is a regular IOU token amount, and can be parsed according to existing rules for those amounts). 
+3. Ignore (for now) the 2nd bit (this is the sign-bit, and is always 1 for both XRP and MPT).
+4. Inspect the 3rd bit. If `0`, then parse as an XRP value per usual. However, if `1`, then parse the remaining `STAmount` bytes as an MPT.
+
+#### Encoding for XRP Values (backward compatible)
+
+```
```
Comment (sappenin) [inline XLS-0033d-multi-purpose-tokens/README.md:1030]: After talking to @shawnxie999 we agreed to adjust this section to be this:

```
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ ┌──────────────────┐ ┌──────────────────────┐ ┌──────────────────┐ ┌──────────────────┐ ┌───────────────────────────────────────────┐ │
│ |        0         | |          1           | |        0         | |                  | |                                           | │
│ |                  | |                      | |                  | |     Reserved     | |                XRP Amount                 | │
│ |  "Not XRP" bit   | | Sign bit (always 1;  | |   "IsMPT" bit    | |     (4 bits)     | |                 (57 bits)                 | │
│ | 0=XRP/MPT; 1=IOU | |      positive)       | |   0=XRP; 1=MPT   | |                  | |                                           | │
│ └──────────────────┘ └──────────────────────┘ └──────────────────┘ └──────────────────┘ └───────────────────────────────────────────┘ │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Once that change is made, I'll merge this PR.

Comment (sappenin) [reply XLS-0033d-multi-purpose-tokens/README.md:864]: LGTM

**Review by sappenin — APPROVED** [review]
